ng client: support for prefix_targets_with_hash when downloading targets

MVrachev opened the pull request:
Fixes #1479

**Description of the changes being introduced by the pull request**:

ng client: support for prefix_targets_with_hash

Add support for prefixing targets with their hashes when downloading or
using HASH.FILENAME.EXT as target names.
The introduction of prefix_targets_with_hash was necessary, because
there are use cases like Warehouse where you could use
consistent_snapshot, but without adding a hash prefix to your targets.

When prefix_targets_with_hash is set to True, target files conforming
the format HASH.FILENAME.EXT will be downloaded from the server, but
they will be saved on the client side without their hash prefixes or
FILENAME.EXT.
This makes sure the client won't understand the usage of
prefix_targets_with_hash.

Still, if you want to use HASH.FILENAME.EXT as target names when
downloading, then additionally you need to provide consistent_snapshot
set to True in your root.json. The reason is that the specification uses
consistent_snapshot for the same purpose:
```
If consistent snapshots are not used (see § 6.2 Consistent snapshots),
then the filename used to download the target file is of the fixed form
FILENAME.EXT (e.g., foobar.tar.gz). Otherwise, the filename is of the
form HASH.FILENAME.EXT
(e.g., c14aeb4ac9f4a8fc0d83d12482b9197452f6adf3eb710e3b1e2b79e8d14cb681.foobar.tar.gz),
where HASH is one of the hashes of the targets file listed in the
targets metadata file found earlier in step § 5.6 Update the targets role.
In either case, the client MUST write the file to non-volatile
storage as FILENAME.EXT.
```

See chapter 5.7.3:
https://theupdateframework.github.io/specification/latest/index.html#fetch-target
By default, prefix_targets_with_hash is set to true to make it easier
to the user to provide uniquely identifiable targets file names by
using consistent_snapshot set to True.

The same behavior of using two flags is used in the legacy code when
calling tuf.client.updater.download_target() in a repository using
prefix_targets_with_hash and consistent_snapshot:
https://github.com/theupdateframework/tuf/blob/develop/tuf/client/updater.py#L1302


Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
ng client: support for prefix_targets_with_hash

Add support for prefixing targets with their hashes when downloading or
using HASH.FILENAME.EXT as target names.
The introduction of prefix_targets_with_hash was necessary, because
there are use cases like Warehouse where you could use
consistent_snapshot, but without adding a hash prefix to your targets.

When prefix_targets_with_hash is set to True, target files conforming
the format HASH.FILENAME.EXT will be downloaded from the server, but
they will be saved on the client side without their hash prefixes or
FILENAME.EXT.
This makes sure the client won't understand the usage of
prefix_targets_with_hash.

Still, if you want to use HASH.FILENAME.EXT as target names when
downloading, then additionally you need to provide consistent_snapshot
set to True in your root.json. The reason is that the specification uses
consistent_snapshot for the same purpose:
"If consistent snapshots are not used (see § 6.2 Consistent snapshots),
then the filename used to download the target file is of the fixed form
FILENAME.EXT (e.g., foobar.tar.gz). Otherwise, the filename is of the
form HASH.FILENAME.EXT
(e.g., c14aeb4ac9f4a8fc0d83d12482b9197452f6adf3eb710e3b1e2b79e8d14cb681.foobar.tar.gz),
where HASH is one of the hashes of the targets file listed in the
targets metadata file found earlier in step § 5.6 Update the targets role.
In either case, the client MUST write the file to non-volatile
storage as FILENAME.EXT."

The same behavior of using two flags is used in the legacy code when
calling tuf.client.updater.download_target() in a repository using
prefix_targets_with_hash and consistent_snapshot.

See chapter 5.7.3:
https://theupdateframework.github.io/specification/latest/index.html#fetch-target
By default, prefix_targets_with_hash is set to true to make it easier
to the user to provide uniquely identifiable targets file names by
using consistent_snapshot set to True.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I will need more time investigating and when I am ready I will update this pr and make it non-draft.

joshuagl commented on tuf/ngclient/updater.py:
```diff
@@ -185,8 +185,10 @@ def download_target(
         target_fileinfo: "TargetFile" = targetinfo["fileinfo"]
+        target_filepath = targetinfo["filepath"]
+        if self.config.consistent_snapshot:
+            target_filepath = target_fileinfo.hashes + "." + target_filepath
```
`TargetFile.hashes` is a dictionary of hash_alg:hash_digest. In the legacy client we currently only download the file prefixed with the first listed hash value in [`_get_target_file()`](https://github.com/theupdateframework/tuf/blob/69dca08b9acdf58b0fa86d7789c6c56aaac1fa78/tuf/client/updater.py#L1261).

The spec describes that we should retrieve the file as _one_ of the hash digest prefixed files (see [5.7.3](https://theupdateframework.github.io/specification/latest/#fetch-target))
> Otherwise, the filename is of the form HASH.FILENAME.EXT (e.g., c14aeb4ac9f4a8fc0d83d12482b9197452f6adf3eb710e3b1e2b79e8d14cb681.foobar.tar.gz), where HASH is one of the hashes of the targets file listed in the targets metadata file found earlier in step

but that the repository should provide copies of the file at each hash digest prefix (see [6.2.1](https://theupdateframework.github.io/specification/latest/#writing-consistent-snapshots))
> consistent target files should be written to non-volatile storage as digest.filename.ext. This means that if the referrer metadata lists N cryptographic hashes of the referred file, then there must be N identical copies of the referred file, where each file will be distinguished only by the value of the digest in its filename.

Therefore the legacy implementation would be appropriate to replicate here – get the first hash from `target_fileinfo.hashes` and prepend it to `target_filepath`.

MVrachev commented:
This pr is ready for review.
I have changed the way I construct the consistent target name.

More importantly, I made sure that when `consistent_snapshot` is True, in `download_target` we are
downloading target files with names conforming to the schema `HASH.FILENAME.EXT`, BUT when
they are saved on the client side they are saved with their non-consistent names or `FILENAME.EXT`.
That's what the current code does when calling `get_target_file`:
https://github.com/theupdateframework/tuf/blob/8e9677d2620c937e3a64e5420e7386f2464f26bd/tuf/client/updater.py#L3183
but the target file itself is saved with the name `destination` which is constructed without the hash prefix:
constructing destination:
https://github.com/theupdateframework/tuf/blob/8e9677d2620c937e3a64e5420e7386f2464f26bd/tuf/client/updater.py#L3161
persisting the temporary file:
https://github.com/theupdateframework/tuf/blob/8e9677d2620c937e3a64e5420e7386f2464f26bd/tuf/client/updater.py#L3186

That way the usage of `consistent_snapshot` doesn't affect the client in any way.

sechkova commented on tuf/ngclient/updater.py:
```diff
@@ -185,8 +185,12 @@ def download_target(
-        target_filepath = targetinfo["filepath"]
         target_fileinfo: "TargetFile" = targetinfo["fileinfo"]
+        target_filepath = targetinfo["filepath"]
+        if self.config.consistent_snapshot:
```
Consistent snapshot is enabled if `self._trusted_set.root.signed.consistent_snapshot` is set to True.
Besides this, the config setting should indicate whether the repository uses hash prefixes or some other naming scheme (in a same way that `prefix_filename_with_hash` is used in the current updater)
Which means that:
1. The config value could be renamed to something similar in meaning to `prefix_filename_with_hash`
2. The condition should become:
`if self._trusted_set.root.signed.consistent_snapshot and self.config.consistent_snapshot`

MVrachev replied:
Good suggestion.
There are cases like PyPi and Warehouse when one could use `consistent_snapshot` without using `prefix_filename_with_hash`.
I will only use the name `prefix_targets_with_hash` because it's clearer that it's meant for targets.

sechkova commented on tuf/ngclient/updater.py:
```diff
@@ -200,6 +204,8 @@ def download_target(
 
             filepath = os.path.join(destination_directory, target_filepath)
+            if self.config.consistent_snapshot:
+                filepath = os.path.join(destination_directory, orig_target_path)
```
Storing `orig_target_path` is unnecessary, you can retrieve it from  `targetinfo["filepath"]` instead.
This way the  second `if self.config.consistent_snapshot` condition becomes redundant too.

MVrachev replied:
Agree, the only downside is that it will require more lines of code than just storing it in `orig_target_path`, but it will make it
easier to understand - one variable less.

sechkova commented on tuf/ngclient/updater.py:
```diff
@@ -185,8 +185,12 @@ def download_target(
+        if self.config.consistent_snapshot:
+            orig_target_path = target_filepath
+            hashes = list(targetinfo["fileinfo"].hashes.values())
+            target_filepath = hashes[0] + "." + target_filepath
```
```suggestion
            target_filepath = f"{hashes[0]}.{target_filepath}"
```

sechkova commented on tuf/ngclient/config.py:
```diff
@@ -15,3 +15,6 @@ class UpdaterConfig:
     targets_max_length: int = 5000000  # bytes
+    # By default, consistent_snapshot is set to true to provide uniquely
+    # identifiable metadata and targets file names for repositories.
+    consistent_snapshot: bool = True
```
`consistent_snapshot` is set by the repository. The client can/should choose whether targets are hash prefixed or not. See comment below.

sechkova commented on tests/test_updater_ng.py:
```diff
@@ -109,14 +110,72 @@ def tearDown(self):
+    """
+
+    delegator_path = os.path.join(self.client_directory, delegator_file)
+    delegator = Metadata.from_file(delegator_path)
```
You can use `sslib_util.get_file_details` to get the hashes, no need of delegator and `Metadata` object.

MVrachev replied:
The problem is that I don't know what hash algorithm to use when calling `sslib.utils.get_file_details()`.
The only one with that knowledge is the delegator.

sechkova replied:
It is not a problem, since you have all the data.

MVrachev replied:
Yep, you are right that it's better to load the data into a dict instead of instantiating an object and doing all the things
around it.

sechkova commented on tuf/ngclient/updater.py:
```diff
@@ -185,8 +185,16 @@ def download_target(
         target_fileinfo: "TargetFile" = targetinfo["fileinfo"]
+        target_filepath = targetinfo["filepath"]
+        consistent_snapshot = False
+        if hasattr(self._trusted_set.root.signed, "consistent_snapshot"):
```
Why is this check needed? "Root" initializes the attribute `consistent_snapshot` in its constrictor?

MVrachev replied:
Yes, you are right this check is incorrect.
I forgot that we are always setting consistent_snapshot even if it's nonexistent in the JSON file itself.

sechkova commented on tuf/ngclient/updater.py:
```diff
@@ -200,6 +208,10 @@ def download_target(
                 ) from e
 
             filepath = os.path.join(destination_directory, target_filepath)
```
```suggestion
            filepath = os.path.join(destination_directory, targetinfo["filepath"])
```

MVrachev replied:
You are right with this suggestion.
That way I will always save the file without the hash prefix - exactly what we want.

sechkova commented:
Please reword the commit.

MVrachev commented:
I squashed my commits and made the following changes:
- found out that I am not setting `consistent_snapshot` to true in any of my tests and I am not actually testing downloading target names in the format `HASH.FILENAME.EXT`. Fixed that by modifying the root bytes, creating a new trusted metadata set and changing the updater internals. Not a great solution, but I think it's better than modifying the client directory used by the rest of the tests.
- reword my commits
- simplified a couple of places thanks to @sechkova suggestions.

sechkova commented on tests/test_updater_ng.py:
```diff
@@ -109,14 +113,96 @@ def tearDown(self):
+
+
+  def test_refresh_on_consistent_targets(self):
+    # Test refresh with consistent snapshot.
```
"on consistent targets", "with consistent snapshot", pick one

sechkova commented on tests/test_updater_ng.py:
```diff
@@ -109,14 +113,96 @@ def tearDown(self):
+    root_path = os.path.join(self.repository_directory, "metadata", "root.json")
+    root = Metadata.from_file(root_path)
+    root.signed.consistent_snapshot = True
+    root_key_path = os.path.join(self.keystore_directory, "root" + '_key')
```
"root" + '_key' = "root_key"

sechkova commented on tests/test_updater_ng.py:
```diff
@@ -109,14 +113,96 @@ def tearDown(self):
+
   def test_refresh(self):
+    # Test refresh without consistent snapshot.
+    self.repository_updater.config.prefix_targets_with_hash = False
```
This config shouldn't be doing anything without consistent_snapshot being set. I'm not sure this explicit disable is needed.

MVrachev commented:
Addressed the new comments made by @sechkova.

sechkova commented on tests/test_updater_ng.py:
```diff
@@ -109,14 +114,94 @@ def tearDown(self):
+      targetname: Name of the target file.
+
+    """
+    delegator_path = os.path.join(self.client_directory, delegator_filename)
```
You keep using this delegator to get the hashes, I keep disagreeing. We need a second opinion.

MVrachev replied:
I am sorry it seemed that I disagree. I am just not sure how can I get information about  hashes that needs to be calculated for a particular delegation if I don't access the delegator.
First, I thought you meant I didn't need to instantiate a Metadata instance in order to solve this problem and that's what I changed, but  it seems like you meant something else.
It looks like I am missing something.
Can you share more how can I simplify this without accessing the delegator?

Commit:
ngclient UpdaterConfig: Add docstring

Docstrings for each class are required by linting tool.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I removed the need of accessing the delegator file by swapping the places of  `get_one_valid_targetinfo()` and `_create_consistent_target()`.
After calling `get_one_valid_targetinfo` I have the information about the target hashes and I can pass it to `_create_consistent_target()` and that's what I did in my last commit.

sechkova commented on tests/test_updater_ng.py:
```diff
@@ -109,14 +114,94 @@ def tearDown(self):
+      target_hash: A string denoting the hash of the target.
+
+    """
+    consistent_target_name = target_hash + "." + targetname
```
```suggestion
    consistent_target_name = f"{target_hash}.{targetname}"
```

sechkova reviewed (commented):
Looks fine to me now.

jku commented on tuf/ngclient/config.py:
```diff
@@ -9,9 +9,16 @@
+    # We need this variable because there are use cases like Warehouse where
+    # you could use consistent_snapshot, but without adding a hash prefix.
+    # By default, prefix_targets_with_hash is set to true to use uniquely
+    # identifiable targets file names for repositories.
```
Let's use the class docstring if this is intended to be documentation: the Attributes sections should work I think?

I think the text is not saying what exactly happens based on the attribute. Maybe something along lines of:
```
    # When consistent snapshots are used, target download URLs are formed by
    # prefixing the filename with a hash digest of file content by default.
    # This can be overridden by setting prefix_targets_with_hash to False.
```
(and "consistent snapshots" should link to the spec explanation of the concept)

jku commented on tests/test_updater_ng.py:
```diff
@@ -109,14 +114,94 @@ def tearDown(self):
+    # This seems better than modifying the actual root file in the client directory
+    # considering that the client directory is used from the rest of the tests.
+    trusted_metadata_set = TrustedMetadataSet(root_modified_bytes)
+    self.repository_updater._trusted_set = trusted_metadata_set
```
I don't like tests modifying internal details.

I'm not sure I understand the comment here: both repository dir and client repo dir are set up in `setUp()` so it should be done for each test.

jku reviewed (commented):
Code part looks good. Could still have a look at variable use but these are nitpicks: 
* why is target_fileinfo declared but then not used four lines later
* should filepath be renamed so it's clear that it's the path for _local_ file

The test code looks a bit bad in that it's adding 80 lines of fairly complex code for a simple test  that doesn't even check that the files it downloaded are the correct ones (half of the code is copy-paste as well)... but fixing this well probably needs a redesign of the whole test architecture so I'm not asking for changes here apart from the comment WRT modifying Updater internals.

MVrachev commented:
I updated the pr so that:
- the test doesn't modify an internal state but instead creates a new root file with `consistent_snapshot` set to True
- changed the `UpdaterConfig` docstring.

MVrachev commented:
> Code part looks good. Could still have a look at variable use but these are nitpicks:
> 
> * why is target_fileinfo declared but then not used four lines later
> * should filepath be renamed so it's clear that it's the path for _local_ file
> 
> The test code looks a bit bad in that it's adding 80 lines of fairly complex code for a simple test that doesn't even check that the files it downloaded are the correct ones (half of the code is copy-paste as well)... but fixing this well probably needs a redesign of the whole test architecture so I'm not asking for changes here apart from the comment WRT modifying Updater internals.

I updated the pr so that:

- the test doesn't modify an internal state but instead creates a new root file with `consistent_snapshot` set to True
- changed the `UpdaterConfig` docstring
- used `targetinfo` where you pointed out
- renamed `filepath` to `local_filepath`
- rebased on top of `develop`

Maybe you can share what do you think could be simplified and we can submit an issue or directly the following pr?

jku commented:
> Maybe you can share what do you think could be simplified and we can submit an issue or directly the following pr?

I (and others) have made some notes here https://github.com/theupdateframework/tuf/issues/1444 -- that one is about generated test data in general but there are comments wrt the other end of the spectrum (vast amounts of metadata that is at least mostly static), maybe no need to file new issues? We'll need to prototype some approaches to see if they make sense...

MVrachev commented:
Yep, I agree with you we need prototypes and more discussions.
I added a test here even though it's not beautiful in order to make sure we don't forget to test the new use cases
we are supporting when implementing new features.

jku reviewed (approved):
Thanks. The UpdaterConfig docs should still be improved but lets not stop this PR with that either... the code looks fine to me.
